File: bindings/java-jni/src/main/java/org/maplibre/nativejni/render/EglContextDescriptor.java
package org.maplibre.nativejni.render;

import java.util.Objects;

/** Mutable EGL context descriptor for OpenGL render targets. */
public final class EglContextDescriptor implements OpenGLContextDescriptor {
  private NativePointer display = NativePointer.NULL;
  private NativePointer config = NativePointer.NULL;
  private NativePointer shareContext = NativePointer.NULL;
  private NativePointer getProcAddress = NativePointer.NULL;

  public EglContextDescriptor() {}

  public EglContextDescriptor(
      NativePointer display, NativePointer config, NativePointer shareContext) {
    this.display = Objects.requireNonNull(display, "display");
    this.config = Objects.requireNonNull(config, "config");
    this.shareContext = Objects.requireNonNull(shareContext, "shareContext");
  }

  public NativePointer display() {
    return display;
  }

  public EglContextDescriptor display(NativePointer display) {
    this.display = Objects.requireNonNull(display, "display");
    return this;
  }

  public NativePointer config() {
    return config;
  }

  public EglContextDescriptor config(NativePointer config) {
    this.config = Objects.requireNonNull(config, "config");
    return this;
  }

  public NativePointer shareContext() {
    return shareContext;
  }

  public EglContextDescriptor shareContext(NativePointer shareContext) {
    this.shareContext = Objects.requireNonNull(shareContext, "shareContext");
    return this;
  }

  public NativePointer getProcAddress() {
    return getProcAddress;
  }

  public EglContextDescriptor getProcAddress(NativePointer getProcAddress) {
    this.getProcAddress = Objects.requireNonNull(getProcAddress, "getProcAddress");
    return this;
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/render/EglContextDescriptor.kt
package org.maplibre.nativeffi.render

/** Mutable EGL context descriptor for OpenGL render targets. */
public class EglContextDescriptor(
  display: NativePointer,
  config: NativePointer,
  shareContext: NativePointer,
  getProcAddress: NativePointer,
  clientApi: OpenGLClientApi = OpenGLClientApi.UNSPECIFIED,
  ownership: OpenGLContextOwnership = OpenGLContextOwnership.SHARED,
) : OpenGLContextDescriptor {
  public var display: NativePointer = display

  public var config: NativePointer = config

  /**
   * Borrowed EGLContext whose share group the session context joins. Required under shared
   * ownership, where the session also takes its client API from this context. A dedicated session
   * joins no share group, so it is null there and names [clientApi] instead.
   */
  public var shareContext: NativePointer = shareContext

  public var getProcAddress: NativePointer = getProcAddress

  /**
   * Client API the session creates its context for. Required under dedicated ownership. A shared
   * session queries [shareContext] for it, so this is ignored there.
   */
  public var clientApi: OpenGLClientApi = clientApi

  public override var ownership: OpenGLContextOwnership = ownership
}
